// @linktr.ee/linkapp/dev-server/shared/theme-presets.ts
/**
 * Theme presets for testing LinkApps in dev mode
 * Based on actual CSS variables from live Linktree profiles
 */

export type FontData = {
  fontFamily: string
  cssUrl?: string
  fontStyle?: string
}

export type ThemePreset = {
  name: string
  variables: Record<string, string>
  font?: FontData
}

export const THEME_PRESETS: Record<string, ThemePreset> = {
  default: {
    name: "Default (Light)",
    variables: {
      "--button-style-text": "#000000",
      "--button-style-background": "#ffffff",
      "--button-style-background-hover":
        "color-mix(in srgb, #ffffff 93%, #000000 7%)",
      "--button-style-border": "none",
      "--button-style-border-color": "transparent",
      "--button-style-shadow": "none",
      "--button-style-shadow-color": "#000000",
      "--button-style-contrast-color": "#000000",
      "--button-style-radius": "8px",
      "--button-style-inner-radius": "min(8px, max(4px, calc(8px - 8px)))",
      "--button-style-skeleton-color": "rgba(0, 0, 0, 0.05)",
    },
    font: {
      fontFamily: "Inter, sans-serif",
      cssUrl: "https://fonts.googleapis.com/css2?family=Inter:wght@400;500;600;700&display=swap",
    },
  },
  dark: {
    name: "Dark",
    variables: {
      "--button-style-text": "#ffffff",
      "--button-style-background": "#1f2937",
      "--button-style-background-hover":
        "color-mix(in srgb, #1f2937 93%, #ffffff 7%)",
      "--button-style-border": "none",
      "--button-style-border-color": "transparent",
      "--button-style-shadow": "none",
      "--button-style-shadow-color": "#000000",
      "--button-style-contrast-color": "#ffffff",
      "--button-style-radius": "8px",
      "--button-style-inner-radius": "min(8px, max(4px, calc(8px - 8px)))",
      "--button-style-skeleton-color": "rgba(255, 255, 255, 0.1)",
    },
    font: {
      fontFamily: "Roboto, sans-serif",
      cssUrl: "https://fonts.googleapis.com/css2?family=Roboto:wght@400;500;700&display=swap",
    },
  },
  purple: {
    name: "Purple",
    variables: {
      "--button-style-text": "#ffffff",
      "--button-style-background": "#7c3aed",
      "--button-style-background-hover":
        "color-mix(in srgb, #7c3aed 93%, #ffffff 7%)",
      "--button-style-border": "none",
      "--button-style-border-color": "transparent",
      "--button-style-shadow": "none",
      "--button-style-shadow-color": "#000000",
      "--button-style-contrast-color": "#ede9fe",
      "--button-style-radius": "16px",
      "--button-style-inner-radius": "min(16px, max(4px, calc(16px - 8px)))",
      "--button-style-skeleton-color": "rgba(124, 58, 237, 0.1)",
    },
    font: {
      fontFamily: "Poppins, sans-serif",
      cssUrl: "https://fonts.googleapis.com/css2?family=Poppins:wght@400;500;600;700&display=swap",
    },
  },
  outline: {
    name: "Outline Style",
    variables: {
      "--button-style-text": "#000000",
      "--button-style-background": "transparent",
      "--button-style-background-hover": "rgba(0, 0, 0, 0.05)",
      "--button-style-border": "2px solid currentColor",
      "--button-style-border-color": "#000000",
      "--button-style-shadow": "none",
      "--button-style-shadow-color": "#000000",
      "--button-style-contrast-color": "#ffffff",
      "--button-style-radius": "24px",
      "--button-style-inner-radius": "min(24px, max(4px, calc(24px - 8px)))",
      "--button-style-skeleton-color": "rgba(0, 0, 0, 0.05)",
    },
    font: {
      fontFamily: "Montserrat, sans-serif",
      cssUrl: "https://fonts.googleapis.com/css2?family=Montserrat:wght@400;500;600;700&display=swap",
    },
  },
  rounded: {
    name: "Fully Rounded",
    variables: {
      "--button-style-text": "#ffffff",
      "--button-style-background": "#059669",
      "--button-style-background-hover":
        "color-mix(in srgb, #059669 93%, #ffffff 7%)",
      "--button-style-border": "none",
      "--button-style-border-color": "transparent",
      "--button-style-shadow": "none",
      "--button-style-shadow-color": "#000000",
      "--button-style-contrast-color": "#d1fae5",
      "--button-style-radius": "28px",
      "--button-style-inner-radius": "28px",
      "--button-style-skeleton-color": "rgba(5, 150, 105, 0.1)",
    },
    font: {
      fontFamily: "Nunito, sans-serif",
      cssUrl: "https://fonts.googleapis.com/css2?family=Nunito:wght@400;500;600;700&display=swap",
    },
  },
};
