// @linktr.ee/linkapp/dev-server/shared/theme-utils.ts
/**
 * Utility functions for theme handling in dev mode
 */

/**
 * Merges theme CSS variables with preview props and creates a theme object
 * compatible with the LinkApp runtime
 */
export function mergeThemeProps(
  themeVariables: Record<string, string>,
  previewProps: Record<string, unknown> = {},
  additionalProps: Record<string, unknown> = {}
) {
  return {
    ...previewProps,
    ...additionalProps,
    theme: {
      cssVariables: themeVariables,
      // Legacy properties (deprecated) - derived from cssVariables
      textColor: themeVariables["--button-style-text"] || "#000000",
      backgroundColor: themeVariables["--button-style-background"] || "#ffffff",
      borderRadius: themeVariables["--button-style-inner-radius"] || "12px",
      borderColor: themeVariables["--button-style-border-color"] || "#e5e7eb",
      backgroundHover: themeVariables["--button-style-background-hover"] || "#f3f4f6",
      contrastColor: themeVariables["--button-style-contrast-color"] || "#ffffff",
      textHoverColor: themeVariables["--linkHoverText"] || "#111827",
    },
  };
}

/**
 * Gets theme variables from URL parameters
 */
export function getThemeFromUrl(presets: Record<string, Record<string, string>>): Record<string, string> {
  const params = new URLSearchParams(window.location.search);
  const themeParam = params.get('theme') || 'default';
  return presets[themeParam] || presets.default;
}
