// @linktr.ee/linkapp/dist/lib/deploy/deploy-utils.js
import pc from 'picocolors';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
// ============================================================================
// CONFIG PATH RESOLUTION
// ============================================================================
/**
 * Resolves the path to linkapp.config.ts in a project.
 */
export function resolveConfigPath(projectPath) {
    const rootConfigPath = join(projectPath, 'linkapp.config.ts');
    const nestedConfigPath = join(projectPath, '.config', 'linkapp.config.ts');
    if (existsSync(rootConfigPath))
        return rootConfigPath;
    if (existsSync(nestedConfigPath))
        return nestedConfigPath;
    throw new Error('Configuration file not found. Expected linkapp.config.ts in project root or .config directory.');
}
/**
 * Extracts the LinkApp ID from a configuration file.
 */
export async function getLinkAppIdFromConfig(configPath) {
    const { createJiti } = await import('jiti');
    const jiti = createJiti(import.meta.url, {
        interopDefault: true,
        moduleCache: false,
    });
    try {
        const config = jiti(configPath);
        if (!config?.manifest?.name) {
            throw new Error('manifest.name is required in linkapp.config.ts');
        }
        return config.manifest.name
            .toLowerCase()
            .replace(/[^a-z0-9]+/g, '-')
            .replace(/^-+|-+$/g, '');
    }
    catch (error) {
        if (error instanceof Error && error.message.includes('manifest.name is required')) {
            throw error;
        }
        throw new Error(`Failed to load linkapp.config.ts: ${error instanceof Error ? error.message : String(error)}`);
    }
}
// ============================================================================
// OUTPUT UTILITIES
// ============================================================================
export function writeLine(message = '') {
    process.stdout.write(`${message}\n`);
}
export function writeSection(title, length = 50) {
    writeLine(pc.bold(pc.cyan(`\n→ ${title}`)));
    writeLine(pc.dim('─'.repeat(length)));
    writeLine();
}
export function writeSeparator(length = 50) {
    writeLine(pc.dim('─'.repeat(length)));
}
// ============================================================================
// CONSTANTS
// ============================================================================
export const DISPLAY = {
    FILE_PREVIEW_LIMIT: 8,
    SEPARATOR_LENGTH: 50,
};
export const TIMESTAMP = {
    SECONDS_TO_MS_THRESHOLD: 10000000000,
};
export const URLS = {
    buildStatus: (linkTypeId, buildId) => `https://linkapp-ci.replit.app/?linkTypeId=${linkTypeId}&buildId=${buildId}`,
    adminCreateLink: {
        production: (linkTypeId) => `https://linktr.ee/admin?action=create-link&linkType=${linkTypeId}`,
        qa: (linkTypeId) => `https://qa.linktr.ee/admin?action=create-link&linkType=${linkTypeId}`,
    },
};
// ============================================================================
// FORMATTERS
// ============================================================================
export function formatTimestamp(timestamp) {
    if (!timestamp)
        return null;
    const timestampMs = timestamp < TIMESTAMP.SECONDS_TO_MS_THRESHOLD ? timestamp * 1000 : timestamp;
    return new Date(timestampMs).toLocaleString();
}
export function formatElapsedTime(ms) {
    return `${(ms / 1000).toFixed(1)}s`;
}
export function displayFileList(files, title) {
    writeLine();
    writeLine(pc.bold(title));
    const filesToShow = files.slice(0, DISPLAY.FILE_PREVIEW_LIMIT);
    for (const file of filesToShow) {
        writeLine(pc.dim(`  • ${file}`));
    }
    if (files.length > DISPLAY.FILE_PREVIEW_LIMIT) {
        writeLine(pc.dim(`  • ... and ${files.length - DISPLAY.FILE_PREVIEW_LIMIT} more files`));
    }
}
export function pluralize(count, singular, plural) {
    const word = count === 1 ? singular : plural ?? `${singular}s`;
    return `${count} ${word}`;
}
// ============================================================================
// ERRORS
// ============================================================================
export class DeploymentError extends Error {
    exitCode;
    userMessage;
    constructor(message, exitCode = 1, userMessage) {
        super(message);
        this.name = 'DeploymentError';
        this.exitCode = exitCode;
        if (userMessage !== undefined) {
            this.userMessage = userMessage;
        }
    }
}
export class AuthenticationError extends DeploymentError {
    environment;
    constructor(environment) {
        const envFlag = environment === 'qa' ? ' --qa' : '';
        super('Not authenticated', 1, `Run: npx @linktr.ee/linkapp login${envFlag}`);
        this.environment = environment;
        this.name = 'AuthenticationError';
    }
}
export class ValidationError extends DeploymentError {
    constructor(message, details) {
        super(message, 1);
        this.name = 'ValidationError';
        if (details) {
            this.userMessage = details.join('\n');
        }
    }
}
export class DeploymentCancelledError extends DeploymentError {
    constructor() {
        super('Deployment cancelled by user', 0);
        this.name = 'DeploymentCancelledError';
    }
}
export function isAPIError(error) {
    return (error !== null &&
        typeof error === 'object' &&
        'statusCode' in error &&
        'messages' in error &&
        Array.isArray(error.messages));
}
