// @linktr.ee/linkapp/dist/lib/deploy/generate-manifest-files.js
import { existsSync, mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { createJiti } from "jiti";
import { detectLayouts } from "../build/detect-layouts.js";
export async function generateManifestFiles(projectPath = process.cwd()) {
    // Check for config in root first, then .config directory
    const rootConfigPath = join(projectPath, "linkapp.config.ts");
    const nestedConfigPath = join(projectPath, ".config", "linkapp.config.ts");
    const configPath = existsSync(rootConfigPath)
        ? rootConfigPath
        : nestedConfigPath;
    if (!existsSync(configPath)) {
        throw new Error("Config file not found: linkapp.config.ts");
    }
    try {
        // Create .linkapp directory for generated files
        const linkappDir = join(projectPath, ".linkapp");
        if (!existsSync(linkappDir)) {
            mkdirSync(linkappDir, { recursive: true });
        }
        // Load the TypeScript config using jiti
        const jiti = createJiti(import.meta.url, {
            interopDefault: true,
            moduleCache: false,
        });
        const config = jiti(configPath);
        // Generate manifest.json
        // Note: The LinkApp ID is derived from kebab-case(name) by the backend
        const manifest = {
            name: config.manifest.name,
            tagline: config.manifest.tagline,
            description: config.manifest.description,
            manifest_version: config.manifest.manifest_version,
            version: config.manifest.version,
            supporting_links: config.manifest.supporting_links,
            category: config.manifest.category,
            ...(config.manifest.categoryV2 !== undefined && {
                categoryV2: config.manifest.categoryV2,
            }),
            ...(config.manifest.categoryV2Order !== undefined && {
                categoryV2Order: config.manifest.categoryV2Order,
            }),
            search_terms: config.manifest.search_terms,
            author: config.manifest.author,
        };
        writeFileSync(join(linkappDir, "manifest.json"), JSON.stringify(manifest, null, 2), "utf-8");
        // Auto-detect featured layout support based on app/featured.tsx presence
        const layoutDetection = detectLayouts(projectPath);
        const hasFeaturedLayout = layoutDetection.layouts.some((layout) => layout.name === "featured");
        // Generate settings.json
        const settings = {
            title: config.settings.title,
            overview: config.settings.overview,
            elements: config.settings.elements,
            // Auto-detected: true if app/featured.tsx exists, false otherwise
            supports_featured_layout: hasFeaturedLayout,
            // Include optional fields if present
            ...(config.settings.uses_url !== undefined && {
                uses_url: config.settings.uses_url,
            }),
            ...(config.settings.has_url !== undefined && {
                has_url: config.settings.has_url,
            }),
            ...(config.settings.icon && { icon: config.settings.icon }),
            ...(config.settings.settings_tab_title && {
                settings_tab_title: config.settings.settings_tab_title,
            }),
            ...(config.settings.setup_instructions && {
                setup_instructions: config.settings.setup_instructions,
            }),
            ...(config.settings.has_settings !== undefined && {
                has_settings: config.settings.has_settings,
            }),
            ...(config.settings.featured_chin_position !== undefined && {
                featured_chin_position: config.settings.featured_chin_position,
            }),
            ...(config.settings.featured_head_allow_unlocked_aspect_ratio !== undefined && {
                featured_head_allow_unlocked_aspect_ratio: config.settings.featured_head_allow_unlocked_aspect_ratio,
            }),
            ...(config.settings.sheet_behavior !== undefined && {
                sheet_behavior: config.settings.sheet_behavior,
            }),
            ...(config.settings.featured_head_click_behavior !== undefined && {
                featured_head_click_behavior: config.settings.featured_head_click_behavior,
            }),
        };
        writeFileSync(join(linkappDir, "settings.json"), JSON.stringify(settings, null, 2), "utf-8");
        // Generate url_match_rules.json if present
        if (config.url_match_rules) {
            writeFileSync(join(linkappDir, "url_match_rules.json"), JSON.stringify(config.url_match_rules, null, 2), "utf-8");
        }
    }
    catch (error) {
        throw new Error(`Failed to generate manifest files: ${error instanceof Error ? error.message : error}`);
    }
}
