// @linktr.ee/linkapp/dist/lib/deploy/output.js
import pc from 'picocolors';
import { writeLine, writeSeparator, writeSuccess, writeError, writeWarning, writeLabelValue, } from '../utils/output.js';
import { formatElapsedTime, formatTimestamp } from '../utils/formatters.js';
import { isAPIError, DeploymentCancelledError, AuthenticationError, } from '../utils/errors.js';
import { URLS } from '../utils/constants.js';
/**
 * Displays the successful deployment results including details and next steps.
 *
 * @param context - Deployment context
 * @param result - Upload result from API
 */
export function displayDeploymentSuccess(context, result) {
    const totalTime = formatElapsedTime(Date.now() - context.startTime);
    writeLine();
    writeSeparator();
    writeSuccess('\nDeployment successful!');
    writeIndented(`Completed in ${totalTime}`);
    writeLine();
    writeLine(pc.bold('Details:'));
    writeLabelValue('ID', result.linkType.linkTypeId);
    writeLabelValue('Status', result.linkType.status);
    const timestamp = formatTimestamp(result.linkType.updatedTimestamp);
    if (timestamp) {
        writeLabelValue('Updated', timestamp);
    }
    displayNextSteps(context, result);
}
/**
 * Displays next steps after successful deployment.
 *
 * @param context - Deployment context
 * @param result - Upload result from API
 */
function displayNextSteps(context, result) {
    writeLine();
    writeLine(pc.bold('Next steps:'));
    // Step 1: Build is happening
    if (result.build?.id) {
        const buildUrl = URLS.buildStatus(result.linkType.linkTypeId, result.build.id);
        writeLine(pc.cyan('  1. Building your LinkApp (this may take a few minutes):'));
        writeIndented(buildUrl, 5);
        writeLine();
    }
    // Step 2: Add to page (only after build completes)
    const createUrl = context.env === 'qa'
        ? URLS.adminCreateLink.qa(result.linkType.linkTypeId)
        : URLS.adminCreateLink.production(result.linkType.linkTypeId);
    writeLine(pc.cyan('  2. Once the build completes, add it to your page:'));
    writeIndented(createUrl, 5);
    writeLine();
}
/**
 * Helper function to write indented text with dimmed styling.
 */
function writeIndented(text, indent = 2) {
    writeLine(pc.dim(' '.repeat(indent) + text));
}
/**
 * Displays deployment error information.
 *
 * @param error - Error that occurred during deployment
 */
export function displayDeploymentError(error) {
    writeLine();
    writeSeparator();
    writeError('\nDeployment failed');
    writeLine();
    // Handle specific error types
    if (error instanceof DeploymentCancelledError) {
        writeWarning('Deployment cancelled');
        writeLine();
        return;
    }
    if (error instanceof AuthenticationError) {
        writeIndented(error.userMessage ?? error.message);
        writeLine();
        return;
    }
    // Handle API errors with detailed information
    if (isAPIError(error)) {
        displayAPIError(error);
        return;
    }
    // Handle generic errors
    if (error instanceof Error) {
        writeIndented(error.message);
    }
    else {
        writeIndented(String(error));
    }
    writeLine();
}
/**
 * Displays detailed API error information.
 *
 * @param error - API error with status and messages
 */
function displayAPIError(error) {
    writeLine(pc.bold('Error Details:'));
    writeLine(pc.red(`  Status: ${error.statusCode}`));
    if (error.messages.length > 0) {
        writeLine();
        writeLine(pc.bold('  Validation Errors:'));
        for (const msg of error.messages) {
            writeLine(pc.red(`    • ${msg}`));
        }
    }
    if (error.responseData?.error) {
        writeLine();
        writeLine(pc.bold('  Error Type:'));
        writeLine(pc.red(`    ${error.responseData.error}`));
    }
    writeLine();
}
