// @linktr.ee/linkapp/dev-server/rsbuild.config.ts
import { defineConfig } from '@rsbuild/core'
import { pluginReact } from '@rsbuild/plugin-react'
import { resolve } from 'path'

export default defineConfig({
  plugins: [pluginReact()],
  source: {
    // Define fallback values - these will be overridden by dev.ts
    define: {
      __PREVIEW_PROPS__: 'undefined',
      __SETTINGS_CONFIG__: 'undefined',
    },
  },
  resolve: {
    alias: {
      '@': resolve(__dirname, '..'),
      '@/components': resolve(__dirname, 'components'),
      '@/lib': resolve(__dirname, 'lib'),
    },
  },
  html: {
    template: resolve(__dirname, 'index.html'),
  },
  server: {
    port: 3000,
    strictPort: true,
  },
  tools: {
    postcss: (opts) => {
      // Support Tailwind CSS v4 via PostCSS
      opts.postcssOptions = opts.postcssOptions || {}
      opts.postcssOptions.plugins = opts.postcssOptions.plugins || []

      // Add Tailwind CSS PostCSS plugin
      try {
        const tailwindPostcss = require('@tailwindcss/postcss')
        opts.postcssOptions.plugins.push(tailwindPostcss())
      } catch {
        // @tailwindcss/postcss not installed
      }

      return opts
    },
  },
})
