// @linktr.ee/linkapp/dev-server/postcss/tailwind-source-fallback.js
import { dirname, relative } from 'node:path'
import postcss from 'postcss'

const DEFAULT_GLOB = '**/*.{ts,tsx,js,jsx,mdx,html}'

function toPosixPath(value) {
  return value.replace(/\\/g, '/')
}

export function createTailwindSourceFallback(options = {}) {
  const { projectRoot, glob = DEFAULT_GLOB } = options

  if (!projectRoot) {
    const noop = () => {
      const postcssPlugin = {
        postcssPlugin: 'linkapp:tailwind-source-fallback:noop',
        Once() {},
      }

      postcssPlugin.postcss = true
      return postcssPlugin
    }

    noop.postcss = true
    noop.__linkappTailwindSource = true
    return noop
  }

  const plugin = () => {
    const postcssPlugin = {
      postcssPlugin: 'linkapp:tailwind-source-fallback',
      Once(root, { result }) {
        const sourceFile = typeof result.opts.from === 'string' ? result.opts.from : null
        if (!sourceFile) {
          return
        }

        const normalizedSourceFile = toPosixPath(sourceFile)
        const normalizedProjectRoot = toPosixPath(
          projectRoot.endsWith('/') ? projectRoot : `${projectRoot}/`
        )

        if (!normalizedSourceFile.startsWith(normalizedProjectRoot)) {
          return
        }

        let hasSourceDirective = false
        root.walkAtRules('source', () => {
          hasSourceDirective = true
          return false
        })

        if (hasSourceDirective) {
          return
        }

        const cssDir = dirname(sourceFile)
        const relativeToRoot = relative(cssDir, projectRoot)
        const baseSegment = relativeToRoot === '' ? '.' : relativeToRoot
        const normalizedBase = toPosixPath(baseSegment)
        const trimmedBase = normalizedBase.replace(/\/$/, '')

        const sourcePath =
          trimmedBase === '.' || trimmedBase === ''
            ? `./${glob}`
            : `${trimmedBase}/${glob}`

        const atRule = postcss.atRule({
          name: 'source',
          params: `"${sourcePath}"`,
        })

        let lastImport
        root.walkAtRules('import', (node) => {
          lastImport = node
        })

        if (lastImport) {
          lastImport.after(atRule)
        } else {
          root.prepend(atRule)
        }
      },
    }

    postcssPlugin.postcss = true
    return postcssPlugin
  }

  plugin.postcss = true
  plugin.__linkappTailwindSource = true

  return plugin
}
