// @linktr.ee/linkapp/dist/lib/utils/constants.js
/**
 * Display constants for formatting output
 */
export const DISPLAY = {
    /** Maximum number of files to show in preview list */
    FILE_PREVIEW_LIMIT: 8,
    /** Length of separator lines */
    SEPARATOR_LENGTH: 50,
    /** Character used for separator lines */
    SEPARATOR_CHAR: '─',
};
/**
 * Timestamp constants for date/time handling
 */
export const TIMESTAMP = {
    /** Threshold for detecting seconds vs milliseconds (year 2100 in seconds) */
    SECONDS_TO_MS_THRESHOLD: 10000000000,
};
/**
 * CLI command templates
 */
export const CLI_COMMANDS = {
    login: (qa) => `npx @linktr.ee/linkapp login${qa ? ' --qa' : ''}`,
    build: () => 'linkapp build',
};
/**
 * URL templates for deployment-related links
 */
export const URLS = {
    /**
     * Build status URL for LinkApp CI
     */
    buildStatus: (linkTypeId, buildId) => `https://linkapp-ci.replit.app/?linkTypeId=${linkTypeId}&buildId=${buildId}`,
    /**
     * Admin URL for creating a link in production
     */
    adminCreateLink: {
        production: (linkTypeId) => `https://linktr.ee/admin?action=create-link&linkType=${linkTypeId}`,
        qa: (linkTypeId) => `https://qa.linktr.ee/admin?action=create-link&linkType=${linkTypeId}`,
    },
};
