// @linktr.ee/linkapp/dist/lib/utils/errors.js
/**
 * Base error class for deployment-related errors.
 */
export class DeploymentError extends Error {
    exitCode;
    userMessage;
    constructor(message, exitCode = 1, userMessage) {
        super(message);
        this.name = 'DeploymentError';
        this.exitCode = exitCode;
        if (userMessage !== undefined) {
            this.userMessage = userMessage;
        }
    }
}
/**
 * Error thrown when user is not authenticated.
 */
export class AuthenticationError extends DeploymentError {
    environment;
    constructor(environment) {
        const envFlag = environment === 'qa' ? ' --qa' : '';
        super('Not authenticated', 1, `Run: npx @linktr.ee/linkapp login${envFlag}`);
        this.environment = environment;
        this.name = 'AuthenticationError';
    }
}
/**
 * Error thrown when project validation fails.
 */
export class ValidationError extends DeploymentError {
    constructor(message, details) {
        super(message, 1);
        this.name = 'ValidationError';
        if (details) {
            this.userMessage = details.join('\n');
        }
    }
}
/**
 * Error thrown when deployment is cancelled by user.
 */
export class DeploymentCancelledError extends DeploymentError {
    constructor() {
        super('Deployment cancelled by user', 0);
        this.name = 'DeploymentCancelledError';
    }
}
/**
 * Type guard to check if an error is an API error.
 */
export function isAPIError(error) {
    return (error !== null &&
        typeof error === 'object' &&
        'statusCode' in error &&
        'messages' in error &&
        Array.isArray(error.messages));
}
/**
 * Error thrown when API request fails.
 */
export class APIRequestError extends DeploymentError {
    apiError;
    constructor(apiError) {
        super(`API request failed with status ${apiError.statusCode}`, 1);
        this.apiError = apiError;
        this.name = 'APIRequestError';
    }
}
