// @linktr.ee/linkapp/dist/lib/utils/output.js
import pc from 'picocolors';
/**
 * Writes a line to stdout with a newline.
 * @param message - The message to write (defaults to empty line)
 */
export function writeLine(message = '') {
    process.stdout.write(`${message}\n`);
}
/**
 * Writes a section header with title and separator line.
 * @param title - The section title
 * @param length - Length of the separator line (default: 50)
 */
export function writeSection(title, length = 50) {
    writeLine(pc.bold(pc.cyan(`\n→ ${title}`)));
    writeLine(pc.dim('─'.repeat(length)));
    writeLine();
}
/**
 * Writes an error message in red with an ✗ symbol.
 * @param message - The error message
 */
export function writeError(message) {
    writeLine(pc.red(pc.bold(`✗ ${message}`)));
}
/**
 * Writes a success message in green with a ✓ symbol.
 * @param message - The success message
 */
export function writeSuccess(message) {
    writeLine(pc.green(pc.bold(`✓ ${message}`)));
}
/**
 * Writes a warning message in yellow with a ⚠ symbol.
 * @param message - The warning message
 */
export function writeWarning(message) {
    writeLine(pc.yellow(pc.bold(`⚠ ${message}`)));
}
/**
 * Writes a horizontal separator line.
 * @param length - Length of the separator (default: 50)
 * @param char - Character to use for separator (default: '─')
 */
export function writeSeparator(length = 50, char = '─') {
    writeLine(pc.dim(char.repeat(length)));
}
/**
 * Writes an indented message with dimmed styling.
 * @param message - The message to write
 * @param indent - Number of spaces to indent (default: 2)
 */
export function writeIndented(message, indent = 2) {
    writeLine(pc.dim(' '.repeat(indent) + message));
}
/**
 * Writes a labeled value pair (e.g., "ID: my-linkapp").
 * @param label - The label text
 * @param value - The value text
 * @param indent - Number of spaces to indent (default: 2)
 */
export function writeLabelValue(label, value, indent = 2) {
    const padding = ' '.repeat(indent);
    writeLine(pc.dim(`${padding}${label}: ${value}`));
}
