// @linktr.ee/linkapp/dist/lib/rsbuild/plugins/copy-public.js
import { cpSync, existsSync } from 'node:fs';
/**
 * Rsbuild plugin that ensures public directory files are copied to dist
 */
export function copyPublicDirectory(options) {
    return {
        name: 'linkapp:copy-public',
        setup(api) {
            api.onAfterBuild(() => {
                const { publicDir, outDir } = options;
                if (!existsSync(publicDir)) {
                    return;
                }
                try {
                    cpSync(publicDir, outDir, {
                        recursive: true,
                        force: true,
                        filter: (src) => {
                            // Skip hidden files and directories
                            const basename = src.split('/').pop() || '';
                            return !basename.startsWith('.');
                        },
                    });
                }
                catch (error) {
                    console.warn('Failed to copy public directory:', error);
                }
            });
        },
    };
}
