// @linktr.ee/linkapp/dist/lib/rsbuild/config-factory.js
import { pluginReact } from '@rsbuild/plugin-react';
import { resolve, join } from 'node:path';
import { assetVersioning } from './plugins/asset-versioning.js';
import { setupRuntime } from '../utils/setup-runtime.js';
import { copyPublicDirectory } from './plugins/copy-public.js';
import { brotliCompression } from './plugins/brotli-compression.js';
export function createRsbuildConfig(options) {
    const { mode, port = 3000, host = 'localhost', open = false, sourcemap = false, profile = false, compress = false } = options;
    const isDev = mode === 'development';
    const projectRoot = process.cwd();
    const runtimeRoot = join(projectRoot, '.linkapp');
    const hashedJsFilename = isDev ? '[name].js' : '[name].[contenthash:8].js';
    const hashedCssFilename = isDev ? '[name].css' : '[name].[contenthash:8].css';
    const hashedSvgFilename = isDev ? '[name].svg' : '[name].[contenthash:8].svg';
    const hashedAssetFilename = isDev ? '[name][ext]' : '[name].[contenthash:8][ext]';
    // Setup runtime files in user's project
    setupRuntime(projectRoot);
    const config = {
        mode,
        root: runtimeRoot,
        plugins: [pluginReact()],
        source: {
            entry: {
                index: join(runtimeRoot, 'main.tsx'),
            },
            define: {},
        },
        resolve: {
            alias: {
                '@': resolve(projectRoot),
            },
            dedupe: ['react', 'react-dom'],
        },
        html: {
            template: join(runtimeRoot, 'index.html'),
        },
        output: {
            distPath: {
                root: join(projectRoot, 'dist'),
                js: '.',
                jsAsync: '.',
                css: '.',
                cssAsync: '.',
                svg: '.',
                font: '.',
                image: '.',
                media: '.',
                assets: '.',
            },
            assetPrefix: './',
            cleanDistPath: true,
            sourceMap: sourcemap
                ? {
                    js: 'hidden-source-map',
                    css: true,
                }
                : false,
            minify: !isDev,
            filename: {
                js: hashedJsFilename,
                css: hashedCssFilename,
                svg: hashedSvgFilename,
                font: hashedAssetFilename,
                image: hashedAssetFilename,
                media: hashedAssetFilename,
                assets: hashedAssetFilename,
            },
            filenameHash: true,
            legalComments: 'none',
            injectStyles: false, // Extract CSS to separate files
        },
        server: {
            port,
            host,
            open,
            headers: {
                'Access-Control-Allow-Origin': '*',
            },
            strictPort: false,
            publicDir: {
                name: join(projectRoot, 'public'),
                copyOnBuild: true,
            },
        },
        tools: {
            postcss: (opts) => {
                // Let Rsbuild load the user's postcss.config.* file for Tailwind v4 support
                // Rsbuild will handle the version compatibility between user's postcss and its compiled version
                opts.postcssOptions = opts.postcssOptions || {};
                opts.sourceMap = sourcemap;
                if (sourcemap) {
                    const currentMap = opts.postcssOptions.map;
                    if (currentMap && typeof currentMap === 'object') {
                        opts.postcssOptions.map = {
                            ...currentMap,
                            inline: false,
                            annotation: false,
                        };
                    }
                    else {
                        opts.postcssOptions.map = {
                            inline: false,
                            annotation: false,
                        };
                    }
                }
                else {
                    opts.postcssOptions.map = false;
                }
                return opts;
            },
        },
        performance: {
            printFileSize: !isDev,
            chunkSplit: {
                strategy: 'split-by-experience',
                override: {
                    cacheGroups: {
                        // Separate React vendor bundle for better caching
                        react: {
                            test: /[\\/]node_modules[\\/](react|react-dom|scheduler)[\\/]/,
                            name: 'react-vendor',
                            priority: 20,
                            chunks: 'all',
                        },
                        // Group other node_modules into vendor bundle
                        vendor: {
                            test: /[\\/]node_modules[\\/]/,
                            name: 'vendor',
                            priority: 10,
                            chunks: 'all',
                            minSize: 10000, // Only create vendor chunk if > 10KB
                        },
                    },
                },
            },
        },
    };
    // Add production-only plugins
    if (!isDev) {
        config.plugins = config.plugins || [];
        config.plugins.push(copyPublicDirectory({
            publicDir: join(projectRoot, 'public'),
            outDir: join(projectRoot, 'dist'),
        }), assetVersioning({
            manifestPath: 'build-manifest.json',
        }));
        // Add Brotli compression if requested
        if (compress) {
            config.plugins.push(brotliCompression({
                quality: 11,
                threshold: 1024,
            }));
        }
        // Add bundle analyzer if requested
        if (profile) {
            console.warn('⚠️  Bundle profiling is not available yet in Rsbuild config. Use --sourcemap flag and analyze with external tools.');
        }
    }
    return config;
}
